File: src/CodexiumMagnus.Core/Configuration/CompositeConfigurationResolver.cs
// filepath: src/CodexiumMagnus.Core/Configuration/CompositeConfigurationResolver.cs
using System.Collections.Generic;
using CodexiumMagnus.Core.Models;

namespace CodexiumMagnus.Core.Configuration
{
    /// <summary>
    /// Resolves configuration across multiple layers.
    /// Expected precedence (highest → lowest):
    ///   Session → User Profile → Corpus → System
    ///
    /// Callers may pass the sources in that order.
    /// We always merge from lowest → highest so higher layers win.
    /// </summary>
    public sealed class CompositeConfigurationResolver
    {
        private readonly IReadOnlyList<IConfigurationSource> _sources;

        public CompositeConfigurationResolver(IReadOnlyList<IConfigurationSource> sources)
        {
            _sources = sources;
        }

        public TypographyConfig GetEffectiveTypography()
        {
            var result = new TypographyConfig();

            // merge from lowest → highest
            for (int i = _sources.Count - 1; i >= 0; i--)
            {
                var layer = _sources[i].GetTypography();
                if (layer is null) continue;

                if (!string.IsNullOrWhiteSpace(layer.BaseFontFamily))
                    result.BaseFontFamily = layer.BaseFontFamily;

                if (layer.BaseFontSizePt.HasValue)
                    result.BaseFontSizePt = layer.BaseFontSizePt;

                if (layer.HeadingScale is not null && layer.HeadingScale.Count > 0)
                    result.HeadingScale = layer.HeadingScale;

                if (layer.PrintOptions is not null)
                    result.PrintOptions = layer.PrintOptions;
            }

            return result;
        }

        public BibliographyConfig GetEffectiveBibliography()
        {
            var result = new BibliographyConfig();

            // merge from lowest → highest
            for (int i = _sources.Count - 1; i >= 0; i--)
            {
                var layer = _sources[i].GetBibliography();
                if (layer is null) continue;

                if (!string.IsNullOrWhiteSpace(layer.Style))
                    result.Style = layer.Style;

                if (!string.IsNullOrWhiteSpace(layer.SortBy))
                    result.SortBy = layer.SortBy;

                if (!string.IsNullOrWhiteSpace(layer.GroupBy))
                    result.GroupBy = layer.GroupBy;
            }

            return result;
        }
    }
}

File: src/codexium-magnus-core/Configuration/CompositeConfigurationResolver.cpp
#include "CompositeConfigurationResolver.h"

namespace CodexiumMagnus::Core::Configuration {

CompositeConfigurationResolver::CompositeConfigurationResolver(
    const QList<ConfigurationSource*>& sources)
    : m_sources(sources) {
}

TypographyConfig CompositeConfigurationResolver::getEffectiveTypography() const {
    TypographyConfig result;

    // Merge from lowest → highest (reverse iteration)
    for (int i = m_sources.size() - 1; i >= 0; --i) {
        auto* layer = m_sources[i];
        if (!layer) continue;

        auto* typography = layer->getTypography();
        if (!typography) continue;

        if (!typography->baseFontFamily.isEmpty()) {
            result.baseFontFamily = typography->baseFontFamily;
        }

        if (typography->baseFontSizePt > 0.0) {
            result.baseFontSizePt = typography->baseFontSizePt;
        }

        if (!typography->headingScale.isEmpty()) {
            result.headingScale = typography->headingScale;
        }

        // Merge print options
        if (typography->printOptions.pageMarginMm > 0.0) {
            result.printOptions.pageMarginMm = typography->printOptions.pageMarginMm;
        }
        // Note: blackOnWhite is a bool, so we check if it was explicitly set
        // For simplicity, we'll just copy it if the source has it set
        result.printOptions.blackOnWhite = typography->printOptions.blackOnWhite;
    }

    return result;
}

BibliographyConfig CompositeConfigurationResolver::getEffectiveBibliography() const {
    BibliographyConfig result;

    // Merge from lowest → highest (reverse iteration)
    for (int i = m_sources.size() - 1; i >= 0; --i) {
        auto* layer = m_sources[i];
        if (!layer) continue;

        auto* bibliography = layer->getBibliography();
        if (!bibliography) continue;

        if (!bibliography->style.isEmpty()) {
            result.style = bibliography->style;
        }

        if (!bibliography->sortBy.isEmpty()) {
            result.sortBy = bibliography->sortBy;
        }

        if (!bibliography->groupBy.isEmpty()) {
            result.groupBy = bibliography->groupBy;
        }
    }

    return result;
}

} // namespace CodexiumMagnus::Core::Configuration